proposed/2022/Multiple-Equivalent-Framework-Support-TFM-As-Aliases.md

```diff
@@ -0,0 +1,220 @@
+| Restore challenges | restore | 2L | |
+| Pack challenges | restore | L | The answer here is likely that first class support won't be available, but having an experience that's intuitive should be the priority.|
+| dotnet.exe challenges | Non-NuGet dotnet.exe commands |  ? | Seems like most of the commands at hand do treat `TargetFramework` as a simple string, but this would need to confirmed. |
+| dotnet list package | dotnet list package | L | The work here would likely follow the work from the the restore side |
+| Visual Studio challenges | Multi targeting, displaying (transitive) dependencies | L(+) | |
+
+
+
+### Build challenges
+
+In PackageReference, NuGet has a contract with the .NET SDK, where NuGet writes the assets file and the .NET SDK consumes it.
+In most scenarios, people use `dotnet restore` and `dotnet build --no-restore` and there are no issues.
+It is not uncommon that people use `nuget.exe restore` due to the fact that they have non-SDK projects in their solution.
+
+Given that NuGet.exe and .NET SDK do not ship together, it is super common that the assets file was generated by a newer/older version of NuGet.exe than the .NET SDK can support. 
+
+Normally, this works. NuGet rarely makes non-additive changes to the assets file.
+
+Sample part of an assets file:
+
+```json
+  "targets": {
+    ".NETFramework,Version=v4.7.2": {},
+    ".NETFramework,Version=v4.8.0": {}
+  },
+
+```
+
+NuGet's assets file is written pivoting on the actual target framework.
+
+To solve this, NuGet would need to:
+
+- Add vNext for the assets file format, version 4. Version 4 would either a dictionary based on the `TargetFramework` value instead of the effective target framework, built by coming `TargetFrameworkMoniker` and `TargetPlatformMoniker`.
+  - This behavior would probably have to be opt-in to avoid breaking all combinations of new `NuGet.exe`, old `.NET SDK` combinations, but we can choose to intentionally not do this to encourage maintaining the 2 versions equivalently.
+- Make a decision what would be a satisfactory behavior for combining newer/older versions of `NuGet.exe` and `.NET SDK`. This is largely just confirming some of the recent decisions.
+
+To solve this, the .NET SDK would need:
+
+- Be able to `build` with both versions 3 and version 4 of the assets file. The NuGet dependencies would naturally flow into the .NET SDK and the work is to just differentiate between the 2 different versions.
+  - The amount of work needed to be done here would likely be affected by the exact shape of the NuGet APIs. These changes may be seamless.
+
+### Restore challenges
+
+In addition to the assets file, restore has a lock file which has a similar schema, that schema would need to be amended. Fortunately when Central Package Management and its transitive pinning was introduced we introduced the concept of a PackagesLockFile version succesfully. We'd just add a version 3.

```

Comment:
The 3 files have different purposes today. 

dgspec -> Internal NuGet generated and NuGet consumed file. Used as a cache to enable working with unloaded VS projects and something that contains all the inputs affecting restore. Think of this as the restore input, a intepretation of the csproj that affects restore.
assets file -> Nuget/SDK build contract. Think of this as the restore output.
packages.lock.json - This is the lock files that's something in addition to the csproj file.


I'm not sure we'd need to unify the files themselves, but where approrpiate our reader/writer code is the same.

Reply:
I know what purpose they serve. I just asked, technically, would it be possible to separate a project state and dependency states?

The way I think about packages is something you load it externally since you won't have it in source-control. And what the packages provide are the package assets that projects consume. That is also a project state. It doesn't need to know that an asset comes from a project or a package. The dependency state is where all your network laden or external resources you refer to. The project state is created from the project but depends on the dependency state to resolve the assets completely.

In this way you could refer to a package, a project or even a project from an external repo which I think is one of the scenarios that people asked. I forgot what issue it is.

Reply:
I think that's orthogonal to this work. 

For the purposes of the project-package duality, you only really care about project file + assets file. You can ignore the lock file as it is about repeatability, and the dgspec is just a summary of what Nuget wants from the project file.

Reply:
I'm resolving this conversation as out of scope for this spec. While I've seen multiple customers assume that lock files can remove the need for a restore, or request nuget consolidate into fewer files, we need to be able to ship this spec's feature in a reasonable time, and trying to expand the scope will make finishing the implement much slower.